features/steps/steps.py

```diff
@@ -1047,6 +1047,63 @@ def systemd_memory_usage_less_than(context, release, mb_limit):
         )
 
 
+@when(
+    "I prepare the local PPAs to upgrade from `{release}` to `{next_release}`"
+)
+def when_i_create_local_ppas(context, release, next_release):
+    if not context.config.build_pr:
+        return
+
+    # We need Kinetic or greater to support zstd when creating the PPAs
+    launch_machine(context, "kinetic", "ppa")
+    when_i_run_command_on_machine(
+        context, "apt-get update", "with sudo", "ppa"
+    )
+    when_i_run_command_on_machine(
+        context, "apt-get install -y aptly", "with sudo", "ppa"
+    )
+    create_local_ppa(context, release)
+    create_local_ppa(context, next_release)
+    repo_line = "deb [trusted=yes] http://{}:8000 {} main".format(
+        context.instances["ppa"].ip, release
+    )
+    repo_file = "/etc/apt/sources.list.d/local-ua.list"
+    when_i_run_shell_command(
+        context, "printf '{}\n' > {}".format(repo_line, repo_file), "with sudo"
+    )
+    when_i_run_command_on_machine(
+        context,
+        "sh -c 'nohup python3 -m http.server --directory ~/.aptly/public > /dev/null 2>&1 &'",  # noqa
```

Comment:
TIL that aptly has an `aptly serve` subcommand that should do exactly this. Did you try it instead of using `http.server`? Nothing bad with it, but as we're using aptly already...